firebase/functions/callable/createUser.ts

```diff
@@ -0,0 +1,147 @@
+import { functions } from '../main'
+import { getAuth, Auth, UserRecord } from 'firebase-admin/auth'
+import { getFirestore, Firestore, FieldValue } from 'firebase-admin/firestore'
+
+const auth: Auth = getAuth()
+const firestore: Firestore = getFirestore()
+
+/**
+ * Error message thrown when user is not authenticated
+ * @param {string} message Error message
+ * @return {Error} Error object
+ */
+class UnauthenticatedError extends Error {
+  type: string
+  /**
+   * @param {string} message Error message
+   */
+  constructor(message: string) {
+    super(message)
+    this.message = message
+    this.type = 'UnauthenticatedError'
+  }
+}
+
+/**
+ * Error message thrown when user is not an admin
+ */
+class NotAnAdminError extends Error {
+  type: string
+  /**
+   * @param {string} message Error message
+   */
+  constructor(message: string) {
+    super(message)
+    this.message = message
+    this.type = 'NotAnAdminError'
+  }
+}
+
+/**
+ * Error message thrown when role is invalid
+ * @param {string} message Error message
+ * @return {Error} Error object
+ */
+class InvalidRoleError extends Error {
+  type: string
+  /**
+   * @param {string} message Error message
+   */
+  constructor(message: string) {
+    super(message)
+    this.message = message
+    this.type = 'InvalidRoleError'
+  }
+}
+
+const roleIsValid = (role: string) => {
+  // To be adapted with your own list of roles
+  const validRoles = ['editor', 'author']
+  return validRoles.includes(role)
+}
+
+export const createUser = functions
+  .region('australia-southeast1')
+  .https.onCall(async (data, context) => {
+    try {
+      // Checking that the user calling the Cloud Function is authenticated
+      if (!context.auth) {
+        throw new UnauthenticatedError(
+          'The user is not authenticated. ' +
+            'Only authenticated Admin users can create new users.'
+        )
+      }
+
+      // Checking that the user calling the Cloud Function is an Admin user
+      // uid of the user calling the Cloud Function
+      const callerUid = context.auth.uid
+      const callerUserRecord: UserRecord = await auth.getUser(callerUid)
+      if (
+        callerUserRecord?.customClaims &&
+        !callerUserRecord.customClaims.admin
+      ) {
+        throw new NotAnAdminError('Only Admin users can create new users.')
+      }
+
+      // Checking that the new user role is valid
+      const role = data.role
+      if (!roleIsValid(role)) {
+        throw new InvalidRoleError(
+          'The "' + role + '" role is not a valid role'
+        )
+      }
+
+      const userCreationRequest = {
+        userDetails: data,
+        status: 'Pending',
+        createdBy: callerUid,
+        createdOn: FieldValue.serverTimestamp()
+      }
+
+      const userCreationRequestRef = await firestore
```

Comment:
Do we do anything with this ref?

Reply:
yea it used later down the line but i will remove this function as it is not relevant to this issue